### ISSUE TITLE
[kws_streaming] Issue on Quantization for STREAM_EXTERNAL_STATE_INFERENCE(ds-cnn,  micro:  no feature extractor)

### ISSUE THREAD

**Issue by wenjingyang** [google-research/google-research#836]
Hi, 
I followed kws_streaming/models/cnn_test.py.  Generate "feature.npz" to prepare calibration_data. 
[feature.zip](https://github.com/google-research/google-research/files/7242587/feature.zip)

I used the code below to convert the stream_state_external model to quantize model. But the final model I got is still float32. Is there anything I did wrong?  
[stream_state_external.zip](https://github.com/google-research/google-research/files/7242640/stream_state_external.zip)

      npzfile = np.load("feature.npz",allow_pickle=True)
      ttt=npzfile.f.arr_0
      calibration_data = prepare_calibration_data(model,100, ttt)
      def representative_dataset(dtype):
        def _representative_dataset_gen():
            for i in range(len(calibration_data)):
                yield [
                    calibration_data[i][0].astype(dtype),  # input audio packet
                    calibration_data[i][1].astype(dtype),  # conv state
                    calibration_data[i][2].astype(dtype)  # flatten state
                    ]
        return _representative_dataset_gen

        with quantize.quantize_scope():
            fd.write(
                utils.model_to_tflite(sess=sess, model_non_stream=model, flags=flags, mode=mode, save_model_path=path_model,
                                optimizations=optimizations,inference_type=tf.int8,experimental_new_quantizer=True,
                                representative_dataset=representative_dataset(np.float32)))

Convert parameters: 

  converter.inference_type = inference_type
  converter.experimental_new_quantizer = experimental_new_quantizer
  converter.experimental_enable_resource_variables = True
  if representative_dataset is not None:
    converter.representative_dataset = representative_dataset

  converter.target_spec.supported_ops = [
      tf.lite.OpsSet.TFLITE_BUILTINS_INT8, tf.lite.OpsSet.SELECT_TF_OPS
  ]
  converter.inference_input_type = tf.int8
  converter.inference_output_type = tf.int8
  converter.allow_custom_ops = True
  if optimizations:
    converter.optimizations = optimizations
  tflite_model = converter.convert()

**Comment by wenjingyang** [comment]
I used latest tf-nightly.

**Comment by rybakov** [comment]
@wenjingyang I do not know what are you going to achieve with your coversion parameters
If you follow [example](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn_test.py#L68) and set converter parameters:
```
  converter.inference_type = tf.int8
  converter.experimental_new_quantizer = True
  converter.experimental_enable_resource_variables = True
  converter.representative_dataset = representative_dataset

  # this will enable audio_spectrogram and mfcc in TFLite
  converter.target_spec.supported_ops = [
      tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS
  ]
  converter.allow_custom_ops = True
  converter.inference_input_type = tf.int8
  converter.inference_output_type = tf.int8
  converter.optimizations = [tf.lite.Optimize.DEFAULT]
```
then you will get a fully quantized model with quantized inputs and outputs.

If you set parameters:
```
  converter.inference_type = tf.int8
  converter.experimental_new_quantizer = True
  converter.experimental_enable_resource_variables = True
  converter.representative_dataset = representative_dataset

  # this will enable audio_spectrogram and mfcc in TFLite
  converter.target_spec.supported_ops = [
      tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS
  ]
  converter.allow_custom_ops = True
  converter.optimizations = [tf.lite.Optimize.DEFAULT]
```
then your model will be partially quantized: all inputs and outputs will be float including input/output states. All inputs/outputs will be quantized and de-quantized inside of the model.

If you create an issue at google-research without mentioning the owner of the project, then the owner will not be notified and there is no guarantee that the issue will be addressed. I would encourage you to mention the project owner in an issue, for example in case of kws_streaming you can use @rybakov, this way I will get notification and will not miss it.

**Comment by wenjingyang** [comment]
@rybakov Thanks for your help. 
My target is to get a fully quantized model. 
I have tried the [example ](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn_test.py) and it works fine. But I can't get any quantization on my [ds-cnn model](https://github.com/google-research/google-research/files/7242640/stream_state_external.zip). 
 I checked the code and found the layer should be quantized  when [cnn model](https://github.com/google-research/google-research/blob/master/kws_streaming/models/cnn.py) is built. So my understanding is that I should use quantized layer when build ds-cnn model. Right?

**Comment by rybakov** [comment]
@wenjingyang quantize.quantize_layer is used for quantization aware training and then it will be used by post-training quantization in tflite.
So, if you are interested in post-training quantization only, then you do not need quantize.quantize_layer with quantization aware training.
Please let me know what error you observe and how to reproduce it.

I would also suggest to try [ds_tc_resnet model](https://github.com/google-research/google-research/blob/master/kws_streaming/models/ds_tc_resnet.py), with training [example](https://github.com/google-research/google-research/blob/624e3196b6fbfa429480027606387ae682b2a207/kws_streaming/experiments/kws_experiments_12_labels.md#ds_tc_resnet---based-on-matchboxnet). It can be faster and more accurate than ds-cnn.

**Comment by wenjingyang** [comment]
@rybakov  Thanks. 
I generated ds_tc_resnet with your example (--preprocess micro). 
My ds_tc_resnet folder:  [ds_tc_resnet](https://github.com/google-research/google-research/files/7248950/ds_tc_resnet.zip)
I only changed the [train/test.py](https://github.com/google-research/google-research/blob/master/kws_streaming/train/test.py) with adding "representative_dataset".  I uploaded my test.py as [test.txt](https://github.com/google-research/google-research/files/7248973/test.txt) since python file can't be uploaded. 
You can reproduce this issue when you run the script below. I have highlighted the difference with your script. 

$CMD_TRAIN \
--data_url '' \
--data_dir $DATA_PATH/ \
--train_dir $MODELS_PATH/ds_tc_resnet/ \
--mel_upper_edge_hertz 7600 \
--how_many_training_steps 4,4,2,2 \
--learning_rate 0.001,0.0005,0.0002,0.0001 \
--window_size_ms 30.0 \
--window_stride_ms 10.0 \
--mel_num_bins 80 \
--dct_num_features 40 \
--resample 0.15 \
--alsologtostderr \
**--train 0 \**
--use_spec_augment 1 \
--time_masks_number 2 \
--time_mask_max_size 25 \
--frequency_masks_number 2 \
--frequency_mask_max_size 7 \
--pick_deterministically 1 \
**--preprocess micro \**
ds_tc_resnet 
--activation 'relu' \
--dropout 0.0 \
--ds_filters '128, 64, 64, 64, 128, 128' \
--ds_repeat '1, 1, 1, 1, 1, 1' \
--ds_residual '0, 1, 1, 1, 0, 0' \
--ds_kernel_size '11, 13, 15, 17, 29, 1' \
--ds_stride '1, 1, 1, 1, 1, 1' \
--ds_dilation '1, 1, 1, 1, 2, 1'


The text.py difference.  
![image](https://user-images.githubusercontent.com/4407779/135206401-4ac1101f-eece-417b-bfb7-c29b5caab81d.png)

**Comment by ctwillson** [comment]
@rybakov @wenjingyang hi,I want to covert a  internal states model to int8 quantized  model like below which can load on tflite micro.However,I have no about how to implement  representative_dataset_gen.Any good idea?
```
with tf.compat.v1.Session() as sess:
  converter = tf.lite.TFLiteConverter.from_saved_model(SAVED_MODEL)
  converter.optimizations = [tf.lite.Optimize.DEFAULT]
  converter.inference_input_type = tf.int8
  converter.inference_output_type = tf.int8
  def representative_dataset_gen():
    for i in range(1000):
      data, _ = audio_processor.get_data(1, i*1, model_settings,
                                         BACKGROUND_FREQUENCY,
                                         BACKGROUND_VOLUME_RANGE,
                                         TIME_SHIFT_MS,
                                         'testing',
                                         sess)
      flattened_data = np.array(data.flatten(), dtype=np.float32).reshape(1, 16000)
      yield [flattened_data]
  converter.representative_dataset = representative_dataset_gen
  tflite_model = converter.convert()
  tflite_model_size = open(MODEL_TFLITE, "wb").write(tflite_model)
  print("Quantized model is %d bytes" % tflite_model_size)

```
